app/main.py

```diff
@@ -1,34 +1,77 @@
+from dataclasses import dataclass
+
+
+@dataclass
 class Deck:
-    def __init__(self, row, column, is_alive=True):
-        pass
+    row: int
+    column: int
+    is_alive: bool = True
+
+    def hit(self) -> None:
+        self.is_alive = False
 
 
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    def __init__(self, start: tuple[int, int], end: tuple[int, int], is_drowned=False) -> None:
+        self.decks = self._create_decks(start, end)
+        self.is_drowned = is_drowned
+
+    def _create_decks(self, start: tuple[int, int], end: tuple[int, int]) -> list[Deck]:
+        decks = []
+        if start[0] == end[0]:
+            for column in range(start[1], end[1] + 1):
+                decks.append(Deck(start[0], column))
+        else:
+            for row in range(start[0], end[0] + 1):
+                decks.append(Deck(row, start[1]))
+
+        return decks
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+
+        return None
 
     def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+        deck = self.get_deck(row, column)
+        if deck and deck.is_alive:
+            deck.hit()
+            self.is_drowned = all(not d.is_alive for d in self.decks)
+            return "Sunk!" if self.is_drowned else "Hit!"
+        return "Miss!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple[tuple[int, int], tuple[int, int]]]):
+        self.field = {}
+        self.ships = []
+
+        for start, end in ships:
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def fire(self, location: tuple[int, int]) -> str:
+        if location in self.field:
+            ship = self.field[location]
+            result = ship.fire(*location)
+            return result
+        return "Miss!"
```

Comment:
The `fire` method in the `Battleship` class should potentially update the field to reflect the result of the shot. Currently, it only returns the result but does not modify the field to indicate a hit or miss visually.